src/main/java/me/shedaniel/fiber2cloth/impl/Fiber2ClothImpl.java

```diff
@@ -1,87 +1,103 @@
+    private final Map<ConfigNode, Function<ConfigNode, AbstractConfigListEntry<?>>> treeEntryMap = Maps.newHashMap();
+    private final ConfigEntryBuilder configEntryBuilder = ConfigEntryBuilder.create();
     private Runnable saveRunnable;
     private Consumer<Screen> afterInitConsumer = screen -> {};
     
     @Deprecated
-    public Fiber2ClothImpl(Screen parentScreen, String modId, ConfigNode node, String title) {
+    public Fiber2ClothImpl(Screen parentScreen, String modId, ConfigBranch node, String title) {
         this.parentScreen = parentScreen;
         this.node = node;
         this.defaultCategoryNode = node;
         this.modId = modId;
         this.title = title;
         initDefaultFunctionMap();
     }
-    
-    public static <T> T cast(Object o, Class<T> clazz) {
-        if (o == null)
-            return null;
-        return clazz.cast(o);
-    }
-    
-    public static String[] splitLine(String s) {
-        if (s == null)
-            return null;
-        return s.split("\n");
-    }
-    
-    public static <T> List<T> list(T[] o) {
-        if (o == null)
-            return Lists.newArrayList();
-        return Lists.newArrayList(o);
-    }
-    
-    public static <T> Optional<String> error(List<Constraint> constraints, Object value, Class<T> clazz) {
+
+    private static <S, T> Optional<String> error(ConfigType<T, S, ?> type, SerializableType<S> constraints, T value) {
         try {
-            T cast = clazz.cast(value);
-            for(Constraint constraint : constraints) {
-                if (!constraint.test(cast))
-                    return Optional.of(I18n.translate("error.fiber2cloth.invaild.value"));
-            }
-        } catch (ClassCastException e) {
-            return Optional.of(I18n.translate("error.fiber2cloth.when.casting"));
+            S v = type.toSerializedType(value);
+            return error(constraints, v);
+        } catch (FiberConversionException e) {
+            return Optional.of(I18n.translate("error.fiber2cloth.when.casting", e.getMessage()));
+        }
+    }
+
+    private static <S> Optional<String> error(SerializableType<S> constraints, S v) {
+        if (!constraints.accepts(v)) {
+            return Optional.of(I18n.translate("error.fiber2cloth.invalid.value", v, constraints));
         }
         return Optional.empty();
     }
-    
+
+    private static List<String> gatherLocalizedLines(String tt) {
+        List<String> lines = new ArrayList<>();
+        if (I18n.hasTranslation(tt)) {
+            lines.add(I18n.translate(tt));
+        }
+        int i = 1;
+        while (I18n.hasTranslation(tt + "[" + i + "]")) {
+            lines.add(I18n.translate(tt + "[" + i + "]"));
+            i++;
+        }
+        return lines;
+    }
+
     @Override
-    public Fiber2Cloth setDefaultCategoryNode(Node defaultCategoryNode) {
+    public Fiber2Cloth setDefaultCategoryNode(ConfigBranch defaultCategoryNode) {
```

Comment:
Since we are breaking api anyways, we should rename the method names to reflect the new names, for example, this should be setDefaultCategoryBranch